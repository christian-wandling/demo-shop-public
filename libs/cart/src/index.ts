export * from './lib/cart.routes';
export * from './lib/cart.facade';
export * from './lib/components/cart/cart.component';
export * from './lib/components/checkout/checkout.component';
export * from './lib/components/cart-icon/cart-icon.component';
