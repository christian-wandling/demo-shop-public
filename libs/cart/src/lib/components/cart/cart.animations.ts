import { animate, style, transition, trigger } from '@angular/animations';

export const animateBackdrop = trigger('animateBackdrop', [
  transition(':enter', [style({ opacity: 0 }), animate('500ms ease-in-out', style({ opacity: 1 }))]),
  transition(':leave', [animate('500ms ease-in-out', style({ opacity: 0 }))]),
]);

export const animateSlideOver = trigger('animateSlideOver', [
  transition(':enter', [
    style({ transform: 'translateX(100%)' }),
    animate('700ms ease-in-out', style({ transform: 'translateX(0)' })),
  ]),
  transition(':leave', [animate('700ms ease-in-out', style({ transform: 'translateX(100%)' }))]),
]);
