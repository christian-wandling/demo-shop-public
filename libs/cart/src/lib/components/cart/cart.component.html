@if (showCart()) {
<dialog
  [@animateBackdrop]
  (keydown.escape)="closeCart()"
  (click)="closeCart()"
  class="relative z-50"
  aria-labelledby="slide-over-title"
  aria-modal="true">
  <div class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity" aria-hidden="true"></div>
  <div class="fixed inset-0 overflow-hidden">
    <div class="absolute inset-0 overflow-hidden"></div>
  </div>
</dialog>
<div [@animateSlideOver] class="z-50 pointer-events-none fixed inset-y-0 right-0 flex max-w-full pl-10">
  <div class="pointer-events-auto w-screen max-w-md">
    <div class="flex h-full flex-col overflow-y-scroll bg-white shadow-xl">
      <div class="flex-1 overflow-y-auto px-4 py-6 sm:px-6">
        <div class="flex items-start justify-between">
          <h2 class="text-lg font-medium text-gray-900" id="slide-over-title">Shopping cart</h2>
          <div class="ml-3 flex h-7 items-center">
            <button (click)="closeCart()" type="button" class="relative -m-2 p-2 text-gray-400 hover:text-gray-500">
              <span class="absolute -inset-0.5"></span>
              <span class="sr-only">Close panel</span>
              <svg
                class="h-6 w-6"
                fill="none"
                viewBox="0 0 24 24"
                stroke-width="1.5"
                stroke="currentColor"
                aria-hidden="true">
                <path stroke-linecap="round" stroke-linejoin="round" d="M6 18L18 6M6 6l12 12" />
              </svg>
            </button>
          </div>
        </div>

        <lib-cart-items [items]="items()" (removeItem)="removeItem($event)" />
      </div>

      <div class="border-t border-gray-200 px-4 py-6 sm:px-6">
        <div class="flex justify-between text-base font-medium text-gray-900">
          <p>Total</p>
          <p>{{ totalPrice() | currency }}</p>
        </div>
        <div class="mt-6">
          <a
            (click)="closeCart()"
            routerLink="/checkout"
            class="flex items-center justify-center rounded-md border border-transparent px-6 py-3 text-base font-medium shadow-sm"
            [ngClass]="{
              'bg-gray-100 text-gray-800 pointer-events-none': !checkoutButtonEnabled(),
              'bg-indigo-600 text-white hover:bg-indigo-700': checkoutButtonEnabled()
            }">
            Checkout
          </a>
        </div>
        <div class="mt-6 flex justify-center text-center text-sm text-gray-500">
          <p>
            or
            <button (click)="closeCart()" type="button" class="font-medium text-indigo-600 hover:text-indigo-500">
              Continue Shopping
              <span aria-hidden="true"> &rarr;</span>
            </button>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
}
