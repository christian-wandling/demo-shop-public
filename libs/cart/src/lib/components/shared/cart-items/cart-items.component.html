<div class="mt-8">
  <div class="flow-root">
    <ul class="-my-6 divide-y divide-gray-200">
      @for (item of items(); track item.id) {
      <li class="flex py-6">
        <div class="h-24 w-24 flex-shrink-0 overflow-hidden rounded-md border border-gray-200">
          <img
            [ngSrc]="item.productThumbnail"
            [height]="96"
            [width]="96"
            [alt]="item.productName"
            onerror="this.src='images/placeholder-image.jpg'" />
        </div>

        <div class="ml-4 flex flex-1 flex-col">
          <div>
            <div class="flex justify-between text-base font-medium text-gray-900">
              <h3>
                <a>{{ item.productName }}</a>
              </h3>
              <p class="ml-4">{{ item.totalPrice | currency }}</p>
            </div>
          </div>
          <div class="flex flex-1 items-end justify-between text-sm">
            <p class="text-gray-500">Qty {{ item.quantity }}</p>

            <div class="flex">
              <button
                (click)="removeItem.emit(item.id)"
                type="button"
                class="font-medium text-indigo-600 hover:text-indigo-500">
                Remove
              </button>
            </div>
          </div>
        </div>
      </li>
      }
    </ul>
  </div>
</div>
