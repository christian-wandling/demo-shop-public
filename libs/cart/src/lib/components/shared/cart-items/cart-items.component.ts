import { ChangeDetectionStrategy, Component, input, output } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { CartItemResponse } from '@demo-shop/api';
import { RouterLink } from '@angular/router';

@Component({
  selector: 'lib-cart-items',
  standalone: true,
  imports: [CommonModule, NgOptimizedImage, RouterLink],
  templateUrl: './cart-items.component.html',
  styleUrl: './cart-items.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CartItemsComponent {
  readonly items = input.required<CartItemResponse[]>();

  removeItem = output<number>();
}
