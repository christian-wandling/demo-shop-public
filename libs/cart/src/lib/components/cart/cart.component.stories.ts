import { applicationConfig, Meta, moduleMetadata, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { provideRouter } from '@angular/router';
import { signal } from '@angular/core';
import { CartItemResponse } from '@demo-shop/api';
import { provideImageLoader } from '@demo-shop/shared';
import { CartComponent } from './cart.component';
import { CartFacade } from '../../cart.facade';
import { CartItemsComponent } from '../shared/cart-items/cart-items.component';
import { provideAnimations } from '@angular/platform-browser/animations';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { mockCartItems } from '../../+mock/mock-cart-items';

const mockCartFacade = (items: CartItemResponse[]) => ({
  removeItem: () => {
    return;
  },
  closeCart: () => {
    return;
  },
  getAll: () => signal(items),
  getTotalPrice: () => {
    const totalPrice = items.reduce((acc, curr) => acc + curr.totalPrice, 0);
    return signal(totalPrice);
  },
  getShowCart: () => signal(true),
  setShowCart: () => {
    return;
  },
});

const meta: Meta<CartComponent> = {
  render: args => ({
    props: args,
    template: `
      <div style="height: 75vh">
        <lib-cart></lib-cart>
      </div>
    `,
  }),
  component: CartComponent,
  title: 'Cart/CartComponent',
  decorators: [
    applicationConfig({
      providers: [provideImageLoader(), provideRouter([{ path: '**', redirectTo: '' }]), provideAnimations()],
    }),
    moduleMetadata({
      imports: [CommonModule, NgOptimizedImage, CartItemsComponent],
    }),
  ],
  argTypes: {
    removeItem: { action: 'removeItem', control: false },
    closeCart: { action: 'closeCart', control: false },
  },
  parameters: {
    controls: {
      exclude: ['#cartFacade', 'checkoutButtonEnabled', 'items', 'totalPrice', 'checkoutButtonEnabled', 'showCart'],
    },
  },
};
export default meta;
type Story = StoryObj<CartComponent>;

export const WithItems: Story = {
  decorators: [
    applicationConfig({
      providers: [{ provide: CartFacade, useValue: mockCartFacade(mockCartItems) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    const productImages = canvas.getAllByRole('img');
    const checkoutButton = canvas.getByText('Checkout');
    expect(productImages).toHaveLength(mockCartItems.length);
    expect(checkoutButton).not.toHaveClass('disabled');
  },
};

export const Empty: Story = {
  decorators: [
    applicationConfig({
      providers: [{ provide: CartFacade, useValue: mockCartFacade([]) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    const checkoutButton = canvas.getByText('Checkout');
    expect(checkoutButton).not.toHaveClass('disabled');
  },
};
