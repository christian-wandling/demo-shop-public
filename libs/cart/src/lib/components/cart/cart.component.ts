import { ChangeDetectionStrategy, Component, computed, inject } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { CartFacade } from '../../cart.facade';
import { RouterLink } from '@angular/router';
import { CartItemsComponent } from '../shared/cart-items/cart-items.component';
import { animateBackdrop, animateSlideOver } from './cart.animations';

@Component({
  selector: 'lib-cart',
  standalone: true,
  imports: [CommonModule, NgOptimizedImage, RouterLink, CartItemsComponent],
  templateUrl: './cart.component.html',
  styleUrl: './cart.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  animations: [animateBackdrop, animateSlideOver],
})
export class CartComponent {
  readonly #cartFacade = inject(CartFacade);
  readonly items = this.#cartFacade.getAll();
  readonly checkoutButtonEnabled = computed(() => this.items().length > 0);
  readonly totalPrice = this.#cartFacade.getTotalPrice();
  readonly showCart = this.#cartFacade.getShowCart();

  removeItem(id: number): void {
    this.#cartFacade.removeItem(id);
  }

  closeCart(): void {
    this.#cartFacade.setShowCart(false);
  }
}
