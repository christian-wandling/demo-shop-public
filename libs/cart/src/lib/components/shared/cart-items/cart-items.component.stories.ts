import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { provideRouter } from '@angular/router';
import { importProvidersFrom } from '@angular/core';
import { CartItemResponse } from '@demo-shop/api';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { faker } from '@faker-js/faker';
import { provideImageLoader } from '@demo-shop/shared';
import { CartItemsComponent } from './cart-items.component';

const items: CartItemResponse[] = Array.from({ length: 5 }).map((_, id) => {
  const quantity = faker.number.int({ min: 1, max: 5 });
  const unitPrice = faker.number.int({ min: 1, max: 20 }) * 10 - 1;

  return {
    id,
    productId: faker.number.int({ min: 1, max: 20 }),
    productName: faker.commerce.productName(),
    productThumbnail: faker.image.urlPicsumPhotos(),
    quantity,
    unitPrice,
    totalPrice: quantity * unitPrice,
  };
});

const meta: Meta<CartItemsComponent> = {
  component: CartItemsComponent,
  title: 'Cart/CartItemsComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule, NgOptimizedImage),
        provideImageLoader(),
        provideRouter([{ path: '**', redirectTo: '' }]),
      ],
    }),
  ],
  argTypes: {
    removeItem: { action: 'removeItem', control: false },
  },
};
export default meta;
type Story = StoryObj<CartItemsComponent>;

export const Default: Story = {
  args: {
    items,
  },
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    const productImages = canvas.getAllByRole('img');
    expect(productImages).toHaveLength(items.length);
  },
};
