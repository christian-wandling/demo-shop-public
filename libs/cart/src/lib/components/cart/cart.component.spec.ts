import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CartComponent } from './cart.component';
import { CartFacade } from '../../cart.facade';
import { Component, Input, signal } from '@angular/core';
import { CartItemDTO } from '@demo-shop/api';
import { provideRouter } from '@angular/router';
import { CartItemsComponent } from '../shared/cart-items/cart-items.component';
import { provideNoopAnimations } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';

describe('CartComponent', () => {
  const showCart = signal(false);

  let component: CartComponent;
  let fixture: ComponentFixture<CartComponent>;
  let cartFacade: CartFacade;

  @Component({ standalone: true, selector: 'lib-cart-items', template: '' })
  class CartItemsStubComponent {
    @Input() items: CartItemDTO[] = [];
  }

  const mockCartItems: CartItemDTO[] = [
    {
      id: 'id',
      productId: 'productId',
      productName: 'productName',
      productThumbnail: 'productThumbnail',
      quantity: 1,
      unitPrice: 1,
      totalPrice: 1,
    },
    {
      id: 'id2',
      productId: 'productId2',
      productName: 'productName2',
      productThumbnail: 'productThumbnail2',
      quantity: 2,
      unitPrice: 1,
      totalPrice: 2,
    },
  ];

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [CartComponent],
      providers: [
        provideRouter([]),
        provideNoopAnimations(),
        {
          provide: CartFacade,
          useValue: {
            getAll: jest.fn().mockReturnValue(signal([mockCartItems])),
            getTotalPrice: jest.fn().mockReturnValue(signal(3)),
            getShowCart: jest.fn().mockReturnValue(showCart),
            removeItem: jest.fn(),
            setShowCart: jest.fn(),
          },
        },
      ],
    })
      .overrideComponent(CartComponent, {
        remove: { imports: [CartItemsComponent] },
        add: { imports: [CartItemsStubComponent] },
      })
      .compileComponents();

    cartFacade = TestBed.inject(CartFacade);
    fixture = TestBed.createComponent(CartComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('showCart', () => {
    it('should not display the component if showCart is set to false', () => {
      expect(component.showCart()).toBe(false);
      expect(fixture.debugElement.query(By.css('lib-cart-items'))).toBeNull();
      expect(fixture).toMatchSnapshot();
    });

    it('should should display the component if showCart is set to true', () => {
      showCart.set(true);
      fixture.detectChanges();

      expect(component.showCart()).toBe(true);
      expect(fixture.debugElement.query(By.css('lib-cart-items'))).toBeTruthy();
      expect(fixture).toMatchSnapshot();
    });
  });

  it('should remove an item', () => {
    const id = '1';

    component.removeItem(id);

    expect(cartFacade.removeItem).toHaveBeenCalledWith(id);
  });

  it('should close the cart', () => {
    component.closeCart();

    expect(cartFacade.setShowCart).toHaveBeenCalledWith(false);
  });
});
