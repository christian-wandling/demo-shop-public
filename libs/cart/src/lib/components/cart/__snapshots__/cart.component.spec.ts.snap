// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartComponent showCart should not display the component if showCart is set to false 1`] = `
<lib-cart
  checkoutButtonEnabled={[Function Function]}
  items={[Function Function]}
  showCart={[Function Function]}
  totalPrice={[Function Function]}
>
</lib-cart>
`;

exports[`CartComponent showCart should should display the component if showCart is set to true 1`] = `
<lib-cart
  checkoutButtonEnabled={[Function Function]}
  items={[Function Function]}
  showCart={[Function Function]}
  totalPrice={[Function Function]}
>
  <dialog
    aria-labelledby="slide-over-title"
    aria-modal="true"
    class="relative z-50 flex ng-trigger ng-trigger-animateBackdrop ng-animating"
  >
    <div
      aria-hidden="true"
      class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
    />
    <div
      class="fixed inset-0 overflow-hidden"
    >
      <div
        class="absolute inset-0 overflow-hidden"
      />
    </div>
  </dialog><div
    class="z-50 pointer-events-none fixed inset-y-0 right-0 flex max-w-full pl-10 ng-trigger ng-trigger-animateSlideOver ng-animating"
  >
    <div
      class="pointer-events-auto w-screen max-w-md"
    >
      <div
        class="flex h-full flex-col bg-white shadow-xl"
      >
        <div
          class="flex-1 overflow-y-auto sm:px-6"
        >
          <div
            class="flex items-start justify-between sticky top-0 py-6 px-4 z-10 border-b-2 border-gray-100 bg-white"
          >
            <h2
              class="text-lg font-medium text-gray-900"
              id="slide-over-title"
            >
              Shopping cart
            </h2>
            <div
              class="ml-3 flex h-7 items-center"
            >
              <button
                class="relative -m-2 p-2 text-gray-400 hover:text-gray-500"
                type="button"
              >
                <span
                  class="absolute -inset-0.5"
                />
                <span
                  class="sr-only"
                >
                  Close panel
                </span>
                <svg
                  aria-hidden="true"
                  class="h-6 w-6"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 18L18 6M6 6l12 12"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </button>
            </div>
          </div>
          <lib-cart-items />
        </div>
        <div
          class="border-t border-gray-200 px-4 py-6 sm:px-6"
        >
          <div
            class="flex justify-between text-base font-medium text-gray-900"
          >
            <p>
              Total
            </p>
            <p>
              $3.00
            </p>
          </div>
          <div
            class="mt-6"
          >
            <a
              class="flex items-center justify-center rounded-md border border-transparent px-6 py-3 text-base font-medium shadow-sm bg-indigo-600 text-white hover:bg-indigo-700"
              href="/checkout"
              routerlink="/checkout"
            >
               Checkout 
            </a>
          </div>
          <div
            class="mt-6 flex justify-center text-center text-sm text-gray-500"
          >
            <p>
               or 
              <button
                class="font-medium text-indigo-600 hover:text-indigo-500"
                type="button"
              >
                 Continue Shopping 
                <span
                  aria-hidden="true"
                >
                   â†’
                </span>
              </button>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</lib-cart>
`;
